### PR TITLE
[Merged by Bors] - feat: run style linters in fix mode and push a commit when "bot fix style" is posted in a PR comment

### PR DESCRIPTION
**Pull request by bryangingechen**
The bot is triggered when writes "bot fix style" on a PR, either in an ordinary comment, a review comment, or a review. It then checks if the author has "write" privileges (or above) and if so, will react with a 🚀 emoji on the triggering comment and then lint and push a commit to the PR branch.

Note that this uses the NIGHTLY_TESTING secret.

See [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/reviewdog/near/460017984).

---

- [ ] depends on: #15051

[Open in Gitpod](https://gitpod.io/from-referrer/)

This can be tested in [my fork of mathlib4](https://github.com/bryangingechen/mathlib4). The main things left to test are permissions and other edge cases. So feel free to open weird issues / PRs to my fork and comment on them to try and break things. I think only people with write access should be able to trigger the bot now, so if you want write access to my fork let me know here or on Zulip!

For starters, if someone would like to try commenting `bot fix style` on https://github.com/bryangingechen/mathlib4/pull/2 (also in a PR review / PR review comment) I can check if the permissions check is working properly. (Since no one other than me has write access at this time, hopefully the bot will ignore your comment.)

### CODE REVIEW

**Comment by bryangingechen** [comment]
Let's make sure nothing explodes if I do this:

bot fix style

**Comment by bryangingechen** [comment]
Ah, it makes sense that the workflow that triggers on comments won't run from a PR.

File: .github/workflows/bot_fix_style_comment.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          set -o pipefail
+          curl -sSfL https://github.com/leanprover/elan/releases/download/v3.1.1/elan-x86_64-unknown-linux-gnu.tar.gz | tar xz
+          ./elan-init -y --default-toolchain none
+          echo "$HOME/.elan/bin" >> "${GITHUB_PATH}"
+
+      # run the same linting step as in lint_and_suggest_pr.yaml
+      - name: lint
+        continue-on-error: true  # allows the following `reviewdog` step to add GitHub suggestions
+        run: |
+          ./scripts/lint-style.sh --fix
```
Comment (jcommelin) [inline .github/workflows/bot_fix_style_comment.yaml:37]: Could you please also make it run `./scripts/lint-bib.sh` and `lake exe mk_all`?

**Comment by adomani** [comment]
> Ah, it makes sense that the workflow that triggers on comments won't run from a PR.

You may still be able to test this by modifying a workflow that reacts to comments that already is in master.

As far as I can tell, if the workflow file is in master, then PRs modifying it work with the *modified* script.  So you could override the `maintainter merge` script to perform what you want in a branch and CI *should* act on it.

**Comment by bryangingechen** [comment]
This will conflict with #15051, so I'll work on this more after that one is merged.

File: Mathlib/Algebra/Quandle.lean [hunk 2]
```diff
@@ -152,7 +152,7 @@ identity holds.
 -/
 lemma act_act_self_eq (x y : S) : (x ◃ y) ◃ x = x ◃ y := by
   have h : (x ◃ y) ◃ x = (x ◃ y) ◃ (x ◃ 1) := by rw [act_one]
-  rw [h, ← Shelf.self_distrib, act_one]
+  rw [h, ←Shelf.self_distrib, act_one]
```
Comment (bryangingechen) [inline Mathlib/Algebra/Quandle.lean:155]: @grunweg Before I rebased this PR on #15051, reviewdog complained about the lack of a space here, so maybe one of the linters was lost in the transition?

Comment (grunweg) [reply Mathlib/Algebra/Quandle.lean:155]: Let me investigate: if I remove this space on master and run `lake exe lint-style` or `lake exe lint-style --fix` locally, I rightfully get an error: `Mathlib/Algebra/Quandle.lean : line 155 : ERR_ARR : Missing space after '←'.`
Thanks looks alright to me. Sorry.

Comment (bryangingechen) [reply Mathlib/Algebra/Quandle.lean:155]: Looks like the error popped up in the run on the latest commit here too, but reviewdog didn't pick it up: https://github.com/leanprover-community/mathlib4/actions/runs/10415147022/job/28845191206?pr=15781#step:6:38

Since reviewdog just looks at git diff, maybe the fix option isn't writing the change to disk? (Haven't checked locally, sorry!)

Comment (grunweg) [reply Mathlib/Algebra/Quandle.lean:155]: Let me check - it should do so (as the writing happens in lint-style.py, which is called just the same).

**Comment by bryangingechen** [comment]
bot fix style

**Comment by bryangingechen** [comment]
Hmm, unfortunately the trick of overwriting the maintainer merge workflow didn't work. The run for the "bot fix style" comment above shows that the workflow on `master` was used: https://github.com/leanprover-community/mathlib4/actions/runs/10460233262/workflow

**Comment by bryangingechen** [comment]
This can be tested in [my fork of mathlib4](https://github.com/bryangingechen/mathlib4). The main things left to test are permissions and other edge cases. So feel free to open weird issues / PRs to my fork and comment on them to try and break things. I think only people with write access should be able to trigger the bot now, so if you want write access to my fork let me know here or on Zulip!

For starters, if someone would like to try commenting `bors fix style` on https://github.com/bryangingechen/mathlib4/pull/2  I can check if the permissions check is working properly. (Since no one other than me has write access at this time, hopefully the bot will ignore your comment.)

File: .github/workflows/bot_fix_style_comment.yaml [hunk 3]
```diff
@@ -7,7 +7,7 @@ on:
 jobs:
   fix_style:
     name: Fix style issues from lint
-    if: github.repository == 'leanprover-community/mathlib4' && (github.event.issue.pull_request != 'null') && github.event.pull_request.draft == false && (startsWith(github.event.comment.body, 'bot fix style') || contains(toJSON(github.event.comment.body), '\nbot fix style'))
+    if: github.repository == 'leanprover-community/mathlib4' && github.event.issue.pull_request && github.event.pull_request.draft == false && (startsWith(github.event.comment.body, 'bot fix style') || contains(toJSON(github.event.comment.body), '\nbot fix style'))
```
Comment (bryangingechen) [inline .github/workflows/bot_fix_style_comment.yaml:10]: reminder to self: this needs to be fixed (in another PR) in the maintainer merge workflows as well

File: .github/workflows/bot_fix_review.yaml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    continue-on-error: true  # do not annoy the user when linting fails, as expected
+    steps:
+      - id: user_permission
+        uses: actions-cool/check-user-permission@v2
+        with:
+          require: 'write'
+
+      - run: echo ${{ steps.user_permission.outputs.require-result }}
+
+      # Maybe no API exists for this yet?
```
Comment (bryangingechen) [inline .github/workflows/bot_fix_review.yaml:22]: Is there a way to add a reaction to a PR review via the API? I tried the following and it gave me a 404 (I'm pretty sure the token I used has the permissions necessary): https://github.com/bryangingechen/mathlib4/actions/runs/10534530644/job/29192315609#step:4:14

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: .github/workflows/bot_fix_style_comment.yaml (hunk 3 above)
Comment (bryangingechen) [reply .github/workflows/bot_fix_style_comment.yaml]: Addressed in #16241
